File: code/ii-service/src/main/java/com/owlunit/service/impl/KeywordServiceImpl.java
package com.owlunit.service.impl;

import com.owlunit.core.orthodoxal.ii.Ii;
import com.owlunit.core.orthodoxal.ii.IiDao;
import com.owlunit.model.cinema.Keyword;
import com.owlunit.service.cinema.KeywordService;
import com.owlunit.service.exception.NotFoundException;
import com.owlunit.service.impl.util.Utils;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

/**
 * @author Anton Chebotaev
 *         Owls Proprietary
 */
public class KeywordServiceImpl implements KeywordService {

    protected IiDao dao;

    private static final String CLASS_MARK_KEY = KeywordServiceImpl.class.getName();
    private static final String CLASS_MARK_VALUE = "#";

    static final String META_KEY_NAME = CLASS_MARK_KEY + ".NAME";

    static Keyword iiToKeyword(IiDao dao, Ii item) {
        Ii meta = Utils.itemWithMeta(dao, item);
        if (!meta.getMeta(CLASS_MARK_KEY).equals(CLASS_MARK_VALUE)) {
            throw new IllegalArgumentException("This is not a keyword");
        }
        return new Keyword(
                item.getId(),
                meta.getMeta(META_KEY_NAME)
        );
    }

    static Ii keywordToIi(IiDao dao, Keyword keyword) throws NotFoundException {
        Ii item = dao.load(keyword.getId());
        if (item == null) {
            throw new NotFoundException(String.format("Keyword(%s)", keyword.getName()));
        } else {
            return item;
        }
    }

    @Override
    public Keyword createKeyword(String name) {
        if (name == null || "".equals(name)) {
            throw new IllegalArgumentException("Can not create keyword with empty or null name");
        }

        Ii item = dao.createInformationItem();
        dao.setMetaUnindexed(item, CLASS_MARK_KEY, CLASS_MARK_VALUE);
        dao.setMeta(item, META_KEY_NAME, name);
        return iiToKeyword(dao, item);
    }

    @Override
    public Keyword loadById(Long id) throws NotFoundException {
        Ii item = dao.load(id);
        if (item == null) {
            throw new NotFoundException(String.format("Keyword(%d)", id));
        } else {
            return iiToKeyword(dao, item);
        }
    }

    @Override
    public Keyword loadByName(String name) throws NotFoundException {
        if (name == null || "".equals(name)) {
            throw new IllegalArgumentException("Can not load keyword with empty or null name");
        }

        Collection<Ii> blankItems = dao.load(META_KEY_NAME, name);
        if (blankItems.isEmpty()) {
            throw new NotFoundException(String.format("Keyword(%s)", name));
        } else {
            Ii meta = dao.loadMeta(blankItems.iterator().next());
            return iiToKeyword(dao, meta);
        }
    }

    @Override
    public Keyword loadOrCreateKeyword(String name) {
        try {
            return loadByName(name);
        } catch (NotFoundException e) {
            return createKeyword(name);
        }
    }

    @Override
    public List<Keyword> listKeywords() {
        Collection<Ii> blankItems = dao.load(CLASS_MARK_KEY, CLASS_MARK_VALUE);
        return iisToKeywords(dao.loadMeta(blankItems));
    }

    @Override
    public Keyword update(Keyword keyword) throws NotFoundException {
        Ii item = keywordToIi(dao, keyword);
        Ii updated = dao.setMeta(item, META_KEY_NAME, keyword.getName());
        return iiToKeyword(dao, updated);
    }

    /*-- - - - - - - - -\
    |   P R I V A T E   |
    \_________________ */

    private List<Keyword> iisToKeywords(Collection<Ii> keywordsIi) {
        List<Keyword> result = new ArrayList<Keyword>();
        for (Ii ii : keywordsIi) {
            result.add(iiToKeyword(dao, ii));
        }
        return result;
    }

    /*-- - - - - - - - - - - - - - - - -\
    |   G E T T E R S  &  S E T T E R S |
    \__________________________________*/

    public IiDao getDao() {
        return dao;
    }

    public void setDao(IiDao dao) {
        this.dao = dao;
    }
}

File: ii-service/src/main/scala/com/owlunit/service/cinema/impl/KeywordServiceImpl.scala
package com.owlunit.service.cinema.impl

import com.owlunit.core.ii.{Ii, IiDao}
import collection.mutable.ListBuffer
import com.owlunit.service.cinema.{KeywordService, KeywordIi}
import com.owlunit.service.cinema.exception.CinemaException

/**
 * @author Anton Chebotaev
 *         Owls Proprietary
 */

object KeywordServiceImpl {

  private[cinema] val MetaKeyPrefix = this.getClass.getName
  private[cinema] val KeySearch = MetaKeyPrefix + ".SEARCH"
  private[cinema] val KeyName = MetaKeyPrefix + ".NAME"

  private[cinema] def extract(dao: IiDao, items: Seq[Ii]): Seq[KeywordIi] = {
    val extracted = items.map(item => extractOne(dao, item))
    for (Some(keyword) <- extracted) yield keyword
  }

  private[cinema] def extractOne(dao: IiDao, item: Ii): Option[KeywordIi] = {
    val meta = withMeta(dao, item)
    meta.metaValue(MetaKeyPrefix) match {
      case None => None
      case Some(_) => Some(new KeywordIi(
        meta.id,
        meta.metaValue(KeywordServiceImpl.KeyName).get
      ))
    }
  }

}

trait KeywordServiceImpl extends KeywordService {
  import KeywordServiceImpl._

  def dao: IiDao

  def createKeyword(name: String) = createKeyword(new KeywordIi(0, name))

  def createKeyword(sample: KeywordIi): KeywordIi = {
    val item = dao.createIi
    dao.setMetaUnindexed(item, MetaKeyPrefix, "#")
    dao.setMeta(item, KeyName, sample.name)
    dao.setMeta(item, KeySearch, sample.name.toLowerCase)
    dao.setMeta(item, CinemaServiceImpl.KeySearch, sample.name.toLowerCase)
    sample.copy(id = item.id)
  }


  def loadKeyword(id: Long): Option[KeywordIi] = extractOne(dao, dao.load(id))

  def loadKeyword(name: String): Option[KeywordIi] = {
    val items = dao.load(KeyName, name)
    items.size match {
      case 0 => None
      case 1 => extractOne(dao, items.iterator.next())
      case _ => throw new CinemaException("Ambiguous load")
    }
  }

  def loadOrCreateKeyword(name: String): KeywordIi = loadKeyword(name) match {
    case Some(x) => x
    case None => createKeyword(new KeywordIi(0, name))
  }

  def searchKeyword(query: String): Seq[KeywordIi] = {
    val items = dao.search(KeySearch, buildQuery(query))
    val iterator = items.iterator
    val result = ListBuffer[KeywordIi]()
    while (iterator.hasNext) {
      val k = extractOne(dao, iterator.next())
      if (k.isDefined)
        result += k.get
    }

    result
  }
}






